content/master/concepts/compositions.md

```diff
@@ -0,0 +1,993 @@
+<!-- vale Google.Quotes = YES -->
+
+```yaml {label="envselect",copy-lines="none"}
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+# Removed for Brevity
+spec:
+  environment:
+    environmentConfigs:
+      - ref:
+          name: example-environment
+  resources:
+  # Removed for Brevity
+```
+
+Inside the {{<hover label="envpatch" line="16">}}patches{{</hover>}} of the 
+resource, use 
+{{<hover label="envpatch" line="17">}}ToEnvironmentFieldPath{{</hover>}} to copy
+data from the resource to the EnvironmentConfig.  
+Use {{<hover label="envpatch" line="20">}}FromEnvironmentFieldPath{{</hover>}}
+to copy data to the resource from the EnvironmentConfig.
+
+```yaml {label="envpatch",copy-lines="none"}
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+# Removed for Brevity
+spec:
+  environment:
+  # Removed for Brevity
+  resources:
+  # Removed for Brevity
+    - name: vpc
+      base:
+        apiVersion: ec2.aws.upbound.io/v1beta1
+        kind: VPC
+        spec:
+          forProvider:
+            cidrBlock: 172.16.0.0/16
+      patches:
+        - type: ToEnvironmentFieldPath
+          fromFieldPath: status.atProvider.id
+          toFieldPath: vpcId
+        - type: FromEnvironmentFieldPath
+          fromFieldPath: tags
+          toFieldPath: spec.forProvider.tags
+```
+
+The [EnvironmentConfigs]({{<ref "./environment-configs" >}}) page has 
+more information on EnvironmentConfigs options and usage.
+
+### Composite functions
+
+Composite functions (`XFNs`) are containers executing any code you define.
+Composite functions can read and write to any resource in the Composition
+they're attached to. This allows composite functions to preform complex
+operations to patch fields, determine if a resource is ready for use or notify
+an external system about the details of resource. 
+
+{{<hint "important" >}}
+Composite functions are an alpha feature. Alpha features aren't enabled by
+default. 
+{{< /hint >}}
+
+To attach a composite function to a Composition define a 
+{{<hover label="xfn" line="7">}}function{{</hover>}} inside the Composition
+{{<hover label="xfn" line="4">}}spec{{</hover>}}.
+
+Provide a {{<hover label="xfn" line="8">}}name{{</hover>}} for the function.   
+The {{<hover label="xfn" line="9">}}type{{</hover>}} is always `Container`.  
+The {{<hover label="xfn" line="11">}}container.image{{</hover>}} is the
+location of the composite function container image.
+
+```yaml {label="xfn",copy-lines="none"}
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+# Removed for Brevity
+spec:
+  resources:
+  # Removed for Brevity
+  functions:
+  - name: rds-instance-password
+    type: Container
+    container:
+      image: xpkg.upbound.io/provider-aws-xfns/random-rds-password:v0.1.0
+```
+
+Read the [composition functions]({{<ref "./composition-functions">}}) page for
+details on building and using composite functions.
+
+
+
+### Storing connection details
```

Comment:
I would expect in this section that we are talking about `connectionDetails` of each child composed resource - that field is important because it specifies what/how to derive connection details to send up the chain from each child.  Is it mentioned anywhere? This section feels like the right place for it...

Reply:
What do you have in mind? I avoid specifics here and redirect to the connection secrets part of the MRs pages since all of the fields come from the managed resource.

Reply:
mainly the `connectionDetails` field for composed resources is what i'm thinking of here - that isn't a general MR thing, it's very specific for child composed resources, and it's critical to getting the connection secrets wired up from top to bottom 😉 - feels like it needs to be somewhere, but i can't find it anywhere when searching around now.

See https://docs.crossplane.io/v1.12/concepts/composition/#composite-resource-connection-secrets as a reference for how it is all wired together.  Each of those parts should be mentioned in their relevant section for the whole thing to work.

Reply:
This section has been updated and a new KB article added to clear up secret stuff. 
[connection details](https://deploy-preview-475--crossplane.netlify.app/knowledge-base/guides/connection-details/)